<a href="https://colab.research.google.com/github/adityapathania22/Spark-With-Python/blob/main/Python_with_Spark.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

(Aditya Pathania
A21002) <br>
[aditya pathania](https://www.linkedin.com/in/aditya-pathania-06b989164/)<br>

# Setting up spark for python

In [ ]:
!apt update > /dev/null

In [ ]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null ## install Java for the system , the above function was found out to update the Jave , i got the error multiple times and figured out that 

In [ ]:
!wget -q http://apache.osuosl.org/spark/spark-3.1.2/spark-3.1.2-bin-hadoop3.2.tgz

In [ ]:
!tar xf spark-3.1.2-bin-hadoop3.2.tgz

In [ ]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.1.2-bin-hadoop3.2"

In [ ]:
!pip install pyspark

In [ ]:
from pyspark.sql import SparkSession
spark=SparkSession.builder.master("local[*]").config('spark.ui.port','4050').getOrCreate()

In [ ]:
ap=spark.sparkContext

# We are done with the installing of java , spark with hadoop , now we will start our work with spark 
## as in python we work with data frames here we have such RDD's , we have the data set , and using that we will create the RDD

#### we have the data set now , we will read the set , and have the rdd created for that 

In [ ]:
r='http://kdd.ics.uci.edu/databases/kddcup99/kddcup.data_10_percent.gz'

## we have read the data file , we will create the RDD now using ap , that is our spark context

In [ ]:
r

'http://kdd.ics.uci.edu/databases/kddcup99/kddcup.data_10_percent.gz'

In [ ]:
rdd=ap.textFile(r)

### we will like to see our RDD , with the same head function , here we use take

In [ ]:
rdd.take(5)

IllegalArgumentException: ignored

### here we got some data error , we will try to fix this up 

In [ ]:
!wget -q http://kdd.ics.uci.edu/databases/kddcup99/kddcup.data_10_percent.gz

In [ ]:
r='./kddcup.data_10_percent.gz'

In [ ]:
rdd=ap.textFile(r)

In [ ]:
rdd.take(5)

['0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,239,486,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,19,19,1.00,0.00,0.05,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,235,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,29,29,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,219,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,39,39,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,217,2032,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,49,49,1.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,normal.']

### the error was that we did not read the file to our system 1st , and directly called the link for data in the varibale , this messed up the data columns , then we used wget to read the data into a variable and then created the RDD

In [ ]:
rdd.count()

494021

# the dataset that we have has the summary as -
## back,buffer_overflow,ftp_write,guess_passwd,imap,ipsweep,land,loadmodule,multihop,neptune,nmap,normal,perl,phf,pod,portsweep,rootkit,satan,smurf,spy,teardrop,warezclient,warezmaster.
##duration: continuous.
##protocol_type: symbolic.
##service: symbolic.
##flag: symbolic.
##src_bytes: continuous.
##dst_bytes: continuous.
##land: symbolic.
##wrong_fragment: continuous.
##urgent: continuous.
##hot: continuous.
##num_failed_logins: continuous.
##logged_in: symbolic.
##num_compromised: continuous.
##root_shell: continuous.
##su_attempted: continuous.
##num_root: continuous.
##num_file_creations: continuous.
##num_shells: continuous.
##num_access_files: continuous.
##num_outbound_cmds: continuous.
##is_host_login: symbolic.
##is_guest_login: symbolic.
##count: continuous.
##srv_count: continuous.
##serror_rate: continuous.
##srv_serror_rate: continuous.
##rerror_rate: continuous.
##srv_rerror_rate: continuous.
##same_srv_rate: continuous.
##diff_srv_rate: continuous.
##srv_diff_host_rate: continuous.
##dst_host_count: continuous.
##dst_host_srv_count: continuous.
##dst_host_same_srv_rate: continuous.
##dst_host_diff_srv_rate: continuous.
##dst_host_same_src_port_rate: continuous.
##dst_host_srv_diff_host_rate: continuous.
##dst_host_serror_rate: continuous.
##dst_host_srv_serror_rate: continuous.
##dst_host_rerror_rate: continuous.
##dst_host_srv_rerror_rate: continuous.

In [ ]:
rdd.countByValue()

defaultdict(int,
            {'0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.': 1,
             '0,tcp,http,SF,239,486,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,19,19,1.00,0.00,0.05,0.00,0.00,0.00,0.00,0.00,normal.': 1,
             '0,tcp,http,SF,235,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,29,29,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.': 1,
             '0,tcp,http,SF,219,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,39,39,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.': 1,
             '0,tcp,http,SF,217,2032,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,49,49,1.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,normal.': 1,
             '0,tcp,http,SF,217,2032,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,59,59,1.00,0.00,0.02,0.00,0.00,0.00,0.00,0.

#### the 1 at the end is the starting  we get for the value , we can also use count by key, which is messed up as per the data 
#### why wait for it let check the key also

In [ ]:
rdd.countByKey()

defaultdict(int,
            {'0': 481671,
             '1': 4602,
             '2': 3486,
             '3': 1157,
             '4': 928,
             '5': 896,
             '6': 367,
             '7': 378,
             '8': 249,
             '9': 287})

### this feature of countbyvalue and countbykey was not clear , but once i reached the key value pair i got to understand it , but was checking this featute while working so keeping it intact here

## we can also use the filter function for the RDD , let's say we want to filter the data that is attack 

In [ ]:
norma=rdd.filter(lambda x:'normal.'in x)

In [ ]:
norma.count()

97278

#### we will be using the set features later in the file , how ever to determine the data for attack , we use the set feature here , we can also use !normnal. as well , which will separate out the attack variables from the list 

In [ ]:
attk=rdd.subtract(norma)

In [ ]:
attk.take(5)

['305,tcp,telnet,SF,1735,2766,0,0,0,3,0,1,2,1,0,0,1,0,0,0,0,0,1,1,0.00,0.00,0.00,0.00,1.00,0.00,0.00,2,4,1.00,0.00,0.50,0.50,0.00,0.00,0.00,0.00,buffer_overflow.',
 '0,tcp,telnet,S0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,5,0.83,1.00,0.00,0.00,0.83,0.33,0.00,5,6,1.00,0.00,0.20,0.33,1.00,0.83,0.00,0.00,neptune.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,123,0.48,0.01,0.48,0.00,0.00,0.00,0.00,0.00,smurf.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,133,0.52,0.01,0.52,0.00,0.00,0.00,0.00,0.00,smurf.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,183,0.72,0.01,0.72,0.00,0.00,0.00,0.00,0.00,smurf.']

In [ ]:
attk.count()

396743

### as we see that other than normal there are other fields as well , we would like to check the other distinct labels of the last feature , we will use the map function for any working in the RDD table .this works same the filter function

In [ ]:
dist=rdd.map(lambda x : x[41]).distinct()

In [ ]:
dist.collect()

['0', ',', '1', '6', '2', '4']

## the collect that we use here is the same as the take function the only diff is that the take gives the response out while the collect , stores it in a memory to work on it both gives the same thing as return the data 

## we were getting some issue , which I figured out that we will need to split the data for that as there is a . which is getting the issue , which i found out that the split was not used here , which did wonder's with the data , and we could not find out the exact data , so the split of data from . with , helped the function to work 

In [ ]:
comsplit=rdd.map(lambda x : x.split(","))
dist2=comsplit.map(lambda x: x[41]).distinct()
dist2.collect()


['normal.',
 'buffer_overflow.',
 'loadmodule.',
 'perl.',
 'neptune.',
 'smurf.',
 'guess_passwd.',
 'pod.',
 'teardrop.',
 'portsweep.',
 'ipsweep.',
 'land.',
 'ftp_write.',
 'back.',
 'imap.',
 'satan.',
 'phf.',
 'nmap.',
 'multihop.',
 'warezmaster.',
 'warezclient.',
 'spy.',
 'rootkit.']

#### we will use function to parse the map and the filter in one function , although we main thing behind using the function is , we will take out the independent feature and copy or tag it as the 1st feature for a record , which will make it easy to read the data , also we will use the map for the comma sep to work in case if dot as we did above 

In [ ]:
def par(line):
  el=line.split(',')
  tag=el[41]
  return (tag,el)
funcsplit=rdd.map(par)

In [ ]:
funcsplit.take(1)

[('normal.',
  ['0',
   'tcp',
   'http',
   'SF',
   '181',
   '5450',
   '0',
   '0',
   '0',
   '0',
   '0',
   '1',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '8',
   '8',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   '1.00',
   '0.00',
   '0.00',
   '9',
   '9',
   '1.00',
   '0.00',
   '0.11',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   'normal.'])]

#### here we saw that the take(1) showed the data in a downwards list , we can use the take feature for 1,2,3,4,5 as well , and store it in a varibale and then use the collect feature to display the results

#### as we earlier did the normal separated values , we will do that again using this functuon this time , to ease out the work 

In [ ]:
def par2(line):
  ele=line.split(",")
  tag1=ele[41]
  return (tag1,ele)
fi=rdd.map(par2)
norm2=fi.filter(lambda x: x[0]=='normal.')
a1=norm2.collect()
normal_using_parfunction=len(a1)
normal_using_parfunction

97278

### wow!!! we finally did it, we could have also used it on above created function

In [ ]:
norm_using_par=funcsplit.filter(lambda x: x[0]=='normal.')
l=norm_using_par.collect()
l1=len(l)
l1

97278

### here we used , .collect to get the data , and then we cannot directly print the data , we will have to give some function as in what we need to do in that , so we needed the collect and then check the len of it to found the count

In [ ]:
service_using_function=comsplit.map(lambda x: x[1]).distinct()
service_using_function.collect()

['tcp', 'udp', 'icmp']

In [ ]:
service1_using_function=comsplit.filter(lambda x: x[1]=='tcp')
l22=service1_using_function.collect()
l2=len(l22)
l2

190065

### As we have some population data , what we do with that is we sample , the derive the sample from the data 

# Sampling

In [ ]:
sample= rdd.sample(False,0.1,2002)
# 0.1 is the fraction of the data we will need , and yes random seed 2002 is my fav number , I keep it that way 

In [ ]:
sample.take(5)

['0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,219,1337,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,39,39,1.00,0.00,0.03,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,217,2032,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,6,6,0.00,0.00,0.00,0.00,1.00,0.00,0.00,49,49,1.00,0.00,0.02,0.00,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,177,1985,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,1,0.00,0.00,0.00,0.00,1.00,0.00,0.00,28,119,1.00,0.00,0.04,0.04,0.00,0.00,0.00,0.00,normal.',
 '0,tcp,http,SF,256,1169,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,4,4,0.00,0.00,0.00,0.00,1.00,0.00,0.00,4,139,1.00,0.00,0.25,0.04,0.00,0.00,0.00,0.00,normal.']

## the sample that we take , will be another RDD as well 

In [ ]:
sample.count()

49506

In [ ]:
print("the total data is ",sample.count(),"out of ",rdd.count())

the total data is  49506 out of  494021


In [ ]:
sample_comma_split=sample.map(lambda x: x.split(","))
sample_normal=sample.filter(lambda x: 'normal.'in x )
print('total normal Interactions are ',l1,'out of ',sample_normal.count())

total normal Interactions are  97278 out of  9699


## ( Correct me If I am wrong on this )
### here is another thing I found out that any function you do , let's say we used the distinct on the above function for that we created another RDD , so every function we use we use another RDD for that , which is actually a good way to work with the data , with out making any changes to the real dataset (RDD) or using much memory for the function ouputs , making an another RDD for that will solve the issue , ( Correct me If I am wrong on this )

## here we are using 2 lines of code to split , taking normal as well , above we did it using a function , down we will use anther function as well , to do both the things 

In [ ]:
a=round((sample_normal.count()/sample.count())*100,2)
a2=round((l1/rdd.count())*100,2)
print('we have',a,'% of normal under sample while the sample % is',a2,'%' )

we have 19.59 % of normal under sample while the sample % is 19.69 %


### well that is pretty awesome sampling done by python , we see that we have equal % of data for normal

## also here we have another benifit of using the take , which stores in the memory itelf , and can be used using take . one of them is take sample, also we will use the split and normal count function in one line

In [ ]:
sample_take=rdd.takeSample(False,40000,2002)
gf=[x.split(",") for x in sample_take if 'normal.' in x ]
normal_sample_take=len(gf)
print('we have',(normal_sample_take/40000)*100,'% of sample')


we have 19.482499999999998 % of sample


#### we have the same sample % here also - so which means we can work with the sample for our predictions or explonatory as well 

# Set operations

#### we have basic set operations like add, subs, mult, subs we have already tried that 

In [ ]:
norma=rdd.filter(lambda x : 'normal.'in x )
attck=rdd.subtract(norma)

In [ ]:
attck.take(5)

['305,tcp,telnet,SF,1735,2766,0,0,0,3,0,1,2,1,0,0,1,0,0,0,0,0,1,1,0.00,0.00,0.00,0.00,1.00,0.00,0.00,2,4,1.00,0.00,0.50,0.50,0.00,0.00,0.00,0.00,buffer_overflow.',
 '0,tcp,telnet,S0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,6,5,0.83,1.00,0.00,0.00,0.83,0.33,0.00,5,6,1.00,0.00,0.20,0.33,1.00,0.83,0.00,0.00,neptune.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,123,0.48,0.01,0.48,0.00,0.00,0.00,0.00,0.00,smurf.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,133,0.52,0.01,0.52,0.00,0.00,0.00,0.00,0.00,smurf.',
 '0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,183,0.72,0.01,0.72,0.00,0.00,0.00,0.00,0.00,smurf.']

In [ ]:
attck.count()

396743

In [ ]:
396743+97278

494021

#### well was just checking if the count was right 😅😅😅😅

### that is the same as the total RDD count

#### we have seen in SQL that , we used a cartesian product and combined tables , we will try and use the same feature here as well for the RDD

### let's take a possibility here , we will do one thing , we will pick 2 features , one is the service x[1],i Believe that has tcp, udp etc , and the one is our independent variable that has normal , buffer_overflow.'loadmodule,perl.neptune... etc , we will separate out the 2 rdd one will contain normal entries and one will contain tcp , then we will combine using cartesian and check the no of features we get that has both the entries , let's begin

In [ ]:
vc=rdd.map(lambda x: x.split(','))
prot=vc.map(lambda x: x[1]=='tcp')
#prot.collect() ### well i tried this , it gave a big list of true and false , so I thought to let it go, and only use it for some function after cartesian

dc=rdd.map(lambda x: x.split(','))
serv=dc.map(lambda x: x[2]=='http_443')
#serv.collect()
cart_prd=prot.cartesian(serv).collect()
print('the toal entries with tcp and http_443 are ',len(cart_prd))

KeyboardInterrupt: ignored

##Your session crashed after using all available RAM. If you are interested in access to high-RAM runtimes, you may want to check out Colab Pro.  so this is the error I get , 
#### let me just go with the distinct features and check the no of possibilities we have with them in the data  

In [ ]:
vc1=rdd.map(lambda x: x.split(','))
prot1=vc1.map(lambda x: x[1]).distinct()
prot1.collect() ### well i tried this , it gave a big list of true and false , so I thought to let it go, and only use it for some function after cartesian

dc1=rdd.map(lambda x: x.split(','))
serv1=dc1.map(lambda x: x[2]).distinct()
serv1.collect()
cart_prd1=prot1.cartesian(serv1).collect()
print('the toal combination for prot and serv are ',len(cart_prd1))

the toal combination for prot and serv are  198


### though it was mentioned by jadianes that cartesian is expensive , well it can be seen above

# Agreegate 
### here we will use 3 diff approaches , 1st is that we will map the normal to the front using a diff approach that we used above , then we will manually use the accumulator to find the normal entries and then count the enteries with the normal values to find out the mean of the duration a person was there in the server the 2nd approach that is the agreegate will solve this problem in one go , let's check out 


In [ ]:
n3=rdd.map(lambda x:x.split(','))
n1=n3.filter(lambda x:x[41]=='normal.')
n1.count()

97278

In [ ]:
n2=n3.filter(lambda x:'normal.' in x )
n2.count()

97278

In [ ]:
a2=n3.filter(lambda x:x[41]!='normal.')
a2.count()

396743

In [ ]:
97278+396743

494021

In [ ]:
norm3=n2.map(lambda x:int(x[0]))
#norm3.collect()
attk2=a2.map(lambda x:int(x[0]))

#### what we did above was we kind of like worked the normal enteries into a dummies the same that we use in the dataframe , now it will be easy for us to count and find the mean , basically using agreegate function of mean , the same we will do here

In [ ]:
nor_duration=norm3.reduce(lambda x,y:x+y)
attk_duration=attk2.reduce(lambda x,y:x+y)
print('normal:',nor_duration,',attack:',attk_duration)

normal: 21075991 ,attack: 2626792


## after this we can calulate the mean of the duration a person was active being a normal and the attack 

In [ ]:
print('the mean of normal duration :',nor_duration/n2.count(),'the mean of attack duration:',attk_duration/a2.count())

the mean of normal duration : 216.65732231336992 the mean of attack duration: 6.620890601724542


### sharing some subject knowledge it is clear that , a attacker will have certain motive , the attacker will stay at the server for some time only , so the function was really helpfull in checking this thing , we will now use a simplified way for using the aggreate , with a variable count and an accumultor , we will use 2 accumulators for that , one will do a sum as done above and one will keep the normal /attack count , that we had to do in a separate window to find the count and find the mean , then we will divide the both to find the mean 

In [ ]:
#n3=rdd.map(lambda x:x.split(','))
#n1=n3.filter(lambda x:x[41]=='normal.')
#a2=n3.filter(lambda x:x[41]!='normal.')
#a2.count()
#norm3=n2.map(lambda x:int(x[0]))
#norm3.collect()
#attk2=a2.map(lambda x:int(x[0]))

# this part will be there , now we have 2 rdd's we will work the agreegate now 

norm_agg=norm3.aggregate((0,0),(lambda acc,value:(acc[0]+value,acc[1]+1)),(lambda acc1,acc2:(acc1[0]+acc2[0],acc1[1]+acc2[1])))
print('the normal mean duration was :',norm_agg[0]/norm_agg[1])




the normal mean duration was : 216.65732231336992


## that worked pretty good, now we see that both the values we calculated for the mean duration are the same

In [ ]:
attk_agg=attk2.aggregate((0,0),(lambda acc,value:(acc[0]+value,acc[1]+1)),(lambda acc1,acc2:(acc1[0]+acc2[0],acc1[1]+acc2[1])))
print('the attack duration mean was :',attk_agg[0]/attk_agg[1])

the attack duration mean was : 6.620890601724542


# Key _value pair RDD's -Dictionary

### above in the sheet we did a parse interaction , in that we used a tag for the rdd ,by putting the value of a feature in the front , that is what we will do here using key value pair

In [ ]:
kv=rdd.map(lambda x:x.split(','))
kvdi=kv.map(lambda x:(x[41],x))
kvdi.take(1)

[('normal.',
  ['0',
   'tcp',
   'http',
   'SF',
   '181',
   '5450',
   '0',
   '0',
   '0',
   '0',
   '0',
   '1',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '0',
   '8',
   '8',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   '1.00',
   '0.00',
   '0.00',
   '9',
   '9',
   '1.00',
   '0.00',
   '0.11',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   '0.00',
   'normal.'])]

#### above we used a reduce , that helped in adding the values , and then we created a separate function 2 create 2 accumulators and then add the values in one of them , now we will do the same using the key value pair accumulators by using reducebykey

In [ ]:
kv_agg=kv.map(lambda x:(x[41],float(x[0])))
kv_agg_dur=kv_agg.reduceByKey(lambda x,y:x+y)### this will provide the duration for each network interaction , simply the 1st accumalator that adds the value
kv_agg_dur.collect()

[('normal.', 21075991.0),
 ('buffer_overflow.', 2751.0),
 ('loadmodule.', 326.0),
 ('perl.', 124.0),
 ('neptune.', 0.0),
 ('smurf.', 0.0),
 ('guess_passwd.', 144.0),
 ('pod.', 0.0),
 ('teardrop.', 0.0),
 ('portsweep.', 1991911.0),
 ('ipsweep.', 43.0),
 ('land.', 0.0),
 ('ftp_write.', 259.0),
 ('back.', 284.0),
 ('imap.', 72.0),
 ('satan.', 64.0),
 ('phf.', 18.0),
 ('nmap.', 0.0),
 ('multihop.', 1288.0),
 ('warezmaster.', 301.0),
 ('warezclient.', 627563.0),
 ('spy.', 636.0),
 ('rootkit.', 1008.0)]

### we can see that the normal above is 21075991.0 , that is the 1st accumulator work , we also had a 2nd accumulator that is nothing but gives the count for the interaction , above we did for just normal , now we do for whole , using key value pair

In [ ]:
#kv=rdd.map(lambda x:x.split(','))
#kvdi=kv.map(lambda x:(x[41],x)) this will remain as now the normal , attack enteries work as a key , clear as we did above as well
cou_key=kvdi.countByKey()


In [ ]:
cou_key

defaultdict(int,
            {'back.': 2203,
             'buffer_overflow.': 30,
             'ftp_write.': 8,
             'guess_passwd.': 53,
             'imap.': 12,
             'ipsweep.': 1247,
             'land.': 21,
             'loadmodule.': 9,
             'multihop.': 7,
             'neptune.': 107201,
             'nmap.': 231,
             'normal.': 97278,
             'perl.': 3,
             'phf.': 4,
             'pod.': 264,
             'portsweep.': 1040,
             'rootkit.': 10,
             'satan.': 1589,
             'smurf.': 280790,
             'spy.': 2,
             'teardrop.': 979,
             'warezclient.': 1020,
             'warezmaster.': 20})

## it is cleat that reduce by key  works on rows , and the counts by key  works on columns

## now we know what to do , we have both the values we can find the mean durations, although i thought on it that we have rdd's and how are we going to divide each element corresponding to the exact entry to get the mean , then it was clear by combine by key , that will keep the values as a key and a value in a pair for each element then collect as map , shows the pair , now it will be easy to do any aggregation be it div , mul or mean  

In [ ]:
com_key=kv_agg.combineByKey((lambda x:(x,1)),
                                (lambda acc,value:(acc[0]+value,acc[1]+1)),
                                (lambda acc1, acc2: (acc1[0]+acc2[0], acc1[1]+acc2[1])))
com_key.collectAsMap()

{'back.': (284.0, 2203),
 'buffer_overflow.': (2751.0, 30),
 'ftp_write.': (259.0, 8),
 'guess_passwd.': (144.0, 53),
 'imap.': (72.0, 12),
 'ipsweep.': (43.0, 1247),
 'land.': (0.0, 21),
 'loadmodule.': (326.0, 9),
 'multihop.': (1288.0, 7),
 'neptune.': (0.0, 107201),
 'nmap.': (0.0, 231),
 'normal.': (21075991.0, 97278),
 'perl.': (124.0, 3),
 'phf.': (18.0, 4),
 'pod.': (0.0, 264),
 'portsweep.': (1991911.0, 1040),
 'rootkit.': (1008.0, 10),
 'satan.': (64.0, 1589),
 'smurf.': (0.0, 280790),
 'spy.': (636.0, 2),
 'teardrop.': (0.0, 979),
 'warezclient.': (627563.0, 1020),
 'warezmaster.': (301.0, 20)}

### we have the key and the values combined now , we will use the key and the value to calculate the mean for all the network int types 

In [ ]:
dur_mean_in=com_key.map(lambda k :(k[0],k[1][0]/k[1][1])).collectAsMap()
dur_mean_in

{'back.': 0.1289151157512483,
 'buffer_overflow.': 91.7,
 'ftp_write.': 32.375,
 'guess_passwd.': 2.7169811320754715,
 'imap.': 6.0,
 'ipsweep.': 0.034482758620689655,
 'land.': 0.0,
 'loadmodule.': 36.22222222222222,
 'multihop.': 184.0,
 'neptune.': 0.0,
 'nmap.': 0.0,
 'normal.': 216.65732231336992,
 'perl.': 41.333333333333336,
 'phf.': 4.5,
 'pod.': 0.0,
 'portsweep.': 1915.2990384615384,
 'rootkit.': 100.8,
 'satan.': 0.040276903713027064,
 'smurf.': 0.0,
 'spy.': 318.0,
 'teardrop.': 0.0,
 'warezclient.': 615.2578431372549,
 'warezmaster.': 15.05}

### printing as sorted values

In [ ]:
for tag in sorted(dur_mean_in,key=dur_mean_in.get,reverse=True):
  print(tag,dur_mean_in[tag])

portsweep. 1915.2990384615384
warezclient. 615.2578431372549
spy. 318.0
normal. 216.65732231336992
multihop. 184.0
rootkit. 100.8
buffer_overflow. 91.7
perl. 41.333333333333336
loadmodule. 36.22222222222222
ftp_write. 32.375
warezmaster. 15.05
imap. 6.0
phf. 4.5
guess_passwd. 2.7169811320754715
back. 0.1289151157512483
satan. 0.040276903713027064
ipsweep. 0.034482758620689655
neptune. 0.0
smurf. 0.0
pod. 0.0
teardrop. 0.0
land. 0.0
nmap. 0.0


# Statistics and EDA

#### now we will get started with the stats and the EDA , that is what is the main work of the data scientist , we now have the clear knowledge of the RDD and the functions that we use in it 

## we will keep only the significant and the relevant variables as per the problem that we have , in which the last varible is the dependent variable , while also we will keep the network tpye , services and protcols 

#### we will be using the parse function , run an enumerator in the vector and move only the variables with the index - 1,2,3,41, as they are either str or written info , while we need numerical for this summary 

In [ ]:
import numpy as np
def parse(line):
  line_split=line.split(',')
  useful_indexes=[1,2,3,41]
  clear_line=[item for i, item in enumerate(line_split) if i not in useful_indexes]
  return np.array([float(x) for x in clear_line])

In [ ]:
df_rdd1=rdd.map(parse)

In [ ]:
df_rdd1.take(1)

[array([0.00e+00, 1.81e+02, 5.45e+03, 0.00e+00, 0.00e+00, 0.00e+00,
        0.00e+00, 0.00e+00, 1.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
        0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
        0.00e+00, 8.00e+00, 8.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
        0.00e+00, 1.00e+00, 0.00e+00, 0.00e+00, 9.00e+00, 9.00e+00,
        1.00e+00, 0.00e+00, 1.10e-01, 0.00e+00, 0.00e+00, 0.00e+00,
        0.00e+00, 0.00e+00])]

In [ ]:
df_rdd1.count()

494021

### now we separated out the numerical and the valueable information for the analysis , let us use the basic stast operations and explore the data more 

In [ ]:
from pyspark.mllib.stat import Statistics as st
from math import sqrt

summary=st.colStats(df_rdd1)

In [ ]:
summary.take(1)

AttributeError: ignored

### on exploring the summary instance , it was found out that this is a function that has instances for the data as mean , min max etc , so we will use it to call such summary items

In [ ]:
summary.mean()

array([4.79793025e+01, 3.02561030e+03, 8.68532425e+02, 4.45325199e-05,
       6.43292492e-03, 1.41694381e-05, 3.45187755e-02, 1.51815409e-04,
       1.48246734e-01, 1.02121165e-02, 1.11331300e-04, 3.64356981e-05,
       1.13517442e-02, 1.08294992e-03, 1.09307094e-04, 1.00805431e-03,
       0.00000000e+00, 0.00000000e+00, 1.38658073e-03, 3.32285690e+02,
       2.92906557e+02, 1.76686659e-01, 1.76608808e-01, 5.74334087e-02,
       5.77189431e-02, 7.91547343e-01, 2.09823874e-02, 2.89968038e-02,
       2.32470778e+02, 1.88665670e+02, 7.53779698e-01, 3.09057307e-02,
       6.01934756e-01, 6.68350131e-03, 1.76753964e-01, 1.76442621e-01,
       5.81176104e-02, 5.74116687e-02])

In [ ]:
summary.mean()[1]

3025.6102959186023

### we can call summary for each and every column by using [0],[1], etc 

### let's try out a good way to call the feature

In [ ]:
print('duration stats')# we are only calling for the 1st column that we have , in the next step we will use enumerator to call all the varibles in one funciton
print(summary.mean()[0])

duration stats
47.97930249928637


### we can also call other summary items as well 

In [ ]:
print('duration detailed summary')
print(summary.mean()[0])
print(sqrt(summary.variance()[0]))
print(summary.max()[0])
print(summary.min()[0])
print(summary.count())

duration detailed summary
47.97930249928637
707.746472305374
58329.0
0.0
494021


In [ ]:
import numpy as np
def parse1(line):
  line_split=line.split(',')
  useful_indexes=[1,2,3,41]
  clear_line=[item for i, item in enumerate(line_split) if i not in useful_indexes]
  return (line_split[41],np.array([float(x) for x in clear_line]))

In [ ]:
df_rdd_label=rdd.map(parse1)

In [ ]:
df_rdd_label.take(1)

[('normal.', array([0.00e+00, 1.81e+02, 5.45e+03, 0.00e+00, 0.00e+00, 0.00e+00,
         0.00e+00, 0.00e+00, 1.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
         0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
         0.00e+00, 8.00e+00, 8.00e+00, 0.00e+00, 0.00e+00, 0.00e+00,
         0.00e+00, 1.00e+00, 0.00e+00, 0.00e+00, 9.00e+00, 9.00e+00,
         1.00e+00, 0.00e+00, 1.10e-01, 0.00e+00, 0.00e+00, 0.00e+00,
         0.00e+00, 0.00e+00]))]

### now we have the lable in the front , what we can do is , we can separate out the portsweep as it has the most entries for attack=1040, put it in the rdd and then print it's summary

In [ ]:
port_df_lable=df_rdd_label.filter(lambda x : 'portsweep.' in x )

In [ ]:
port_df_lable.count()

1040

## the no comes out to be same , let's work with the summary

##### the main problem to find the summary is that we have a key/tag/label at the start , so what we do is , we will use the .values and then find the summary

In [ ]:
summary_port=st.colStats(port_df_lable.values())

In [ ]:
summary_port.mean()

array([1.91529904e+03, 6.66707437e+05, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 1.92307692e-03, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 1.82894231e+01,
       1.14519231e+00, 1.87884615e-02, 2.32500000e-02, 9.72076923e-01,
       9.68096154e-01, 8.23211538e-01, 2.14403846e-01, 6.44230769e-04,
       2.49952885e+02, 1.53557692e+00, 4.14423077e-03, 7.64692308e-01,
       8.87326923e-01, 5.76923077e-05, 1.16634615e-02, 2.32500000e-02,
       8.90173077e-01, 9.65259615e-01])

### we see that the 1st column is for the duration , we can check the duration summary values , we know about the mean and the count , so that will be easy 

In [ ]:
print('duration sts summary for portsweep ')
print('mean',summary_port.mean()[0])
print('s.d',sqrt(summary_port.variance()[0]))
print('max',summary_port.max()[0])
print('min',summary_port.min()[0])
print('count',summary.count())## count gives the count for all the data 

duration sts summary for portsweep 
mean 1915.299038461538
s.d 7285.125158537256
max 42448.0
min 0.0
count 494021


### there is also another way that we can use to pass the function as per the lable , we used a key value pair above , we can simply filter the label /tag and then pass the function , after that pass the label and print the summary , let's try out

In [ ]:
def parse1(line):
  line_split=line.split(',')
  useful_indexes=[1,2,3,41]
  clear_line=[item for i, item in enumerate(line_split) if i not in useful_indexes]
  return (line_split[41],np.array([float(x) for x in clear_line]))
### we did this above 



In [ ]:
def sumry(rdd,label):
  lab_df=rdd.map(parse1).filter(lambda x:x[0]==label)
  return st.colStats(lab_df.values())

In [ ]:
por_sum=sumry(rdd,'portsweep.')

In [ ]:
por_sum.mean()

array([1.91529904e+03, 6.66707437e+05, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 1.92307692e-03, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 0.00000000e+00,
       0.00000000e+00, 0.00000000e+00, 0.00000000e+00, 1.82894231e+01,
       1.14519231e+00, 1.87884615e-02, 2.32500000e-02, 9.72076923e-01,
       9.68096154e-01, 8.23211538e-01, 2.14403846e-01, 6.44230769e-04,
       2.49952885e+02, 1.53557692e+00, 4.14423077e-03, 7.64692308e-01,
       8.87326923e-01, 5.76923077e-05, 1.16634615e-02, 2.32500000e-02,
       8.90173077e-01, 9.65259615e-01])

In [ ]:
print('duration summary for portsweep:')
print('mean duration',por_sum.mean()[0])
print('deviation for duration',sqrt(por_sum.variance()[0]))
print('max duration ',por_sum.max()[0])
print('min duration',por_sum.min()[0])


duration summary for portsweep:
mean duration 1915.299038461538
deviation for duration 7285.125158537256
max duration  42448.0
min duration 0.0


### works good , we took a function that gives a lable , then separates out lable , do the summary and print it with the label 

### but this working one on one will take a lot of time and memory , we will need a better table that shows all the summary for all the interaction types , we will get the list of interaction types ,as we have seen above as well , we have around 21 no of list for that , we will print a table with the summary for all

In [ ]:
taking_out_list=rdd.map(lambda x: x.split(','))
listlabel=taking_out_list.map(lambda x:x[41]).distinct()
labels1=listlabel.collect()

In [ ]:
labels1

['normal.',
 'buffer_overflow.',
 'loadmodule.',
 'perl.',
 'neptune.',
 'smurf.',
 'guess_passwd.',
 'pod.',
 'teardrop.',
 'portsweep.',
 'ipsweep.',
 'land.',
 'ftp_write.',
 'back.',
 'imap.',
 'satan.',
 'phf.',
 'nmap.',
 'multihop.',
 'warezmaster.',
 'warezclient.',
 'spy.',
 'rootkit.']

In [ ]:
label_list=['normal.',
 'buffer_overflow.',
 'loadmodule.',
 'perl.',
 'neptune.',
 'smurf.',
 'guess_passwd.',
 'pod.',
 'teardrop.',
 'portsweep.',
 'ipsweep.',
 'land.',
 'ftp_write.',
 'back.',
 'imap.',
 'satan.',
 'phf.',
 'nmap.',
 'multihop.',
 'warezmaster.',
 'warezclient.',
 'spy.',
 'rootkit.']

In [ ]:
sum_by_label=[(label,sumry(rdd,label))for label in label_list]

In [ ]:
sumbylabel_by_duration=[(stat[0],stat[1].mean()[0],sqrt(stat[1].variance()[0]),stat[1].max()[0],stat[1].min()[0])for stat in sum_by_label]

In [ ]:
sumbylabel_by_duration

[('normal.', 216.65732231336938, 1359.213468917662, 58329.0, 0.0),
 ('buffer_overflow.', 91.70000000000003, 97.51468501258388, 321.0, 0.0),
 ('loadmodule.', 36.22222222222222, 41.40886915196362, 103.0, 0.0),
 ('perl.', 41.333333333333336, 14.843629385474879, 54.0, 25.0),
 ('neptune.', 0.0, 0.0, 0.0, 0.0),
 ('smurf.', 0.0, 0.0, 0.0, 0.0),
 ('guess_passwd.', 2.7169811320754715, 11.879810537356365, 60.0, 0.0),
 ('pod.', 0.0, 0.0, 0.0, 0.0),
 ('teardrop.', 0.0, 0.0, 0.0, 0.0),
 ('portsweep.', 1915.299038461538, 7285.125158537256, 42448.0, 0.0),
 ('ipsweep.', 0.034482758620689655, 0.4384391926659592, 7.0, 0.0),
 ('land.', 0.0, 0.0, 0.0, 0.0),
 ('ftp_write.', 32.375, 47.44903280664121, 134.0, 0.0),
 ('back.', 0.1289151157512483, 1.1100621667887005, 14.0, 0.0),
 ('imap.', 6.0, 14.174240399721281, 41.0, 0.0),
 ('satan.', 0.04027690371302706, 0.5224327166508457, 11.0, 0.0),
 ('phf.', 4.5, 5.744562646538029, 12.0, 0.0),
 ('nmap.', 0.0, 0.0, 0.0, 0.0),
 ('multihop.', 184.00000000000003, 253.85100

## we will use our best library pandas to print out the data frame for the values we have

In [ ]:
import pandas as pd
df=pd.DataFrame(sumbylabel_by_duration,columns=['name','mean','standard deviation','max','min'])

In [ ]:
df

name         mean  standard deviation      max    min
0            normal.   216.657322         1359.213469  58329.0    0.0
1   buffer_overflow.    91.700000           97.514685    321.0    0.0
2        loadmodule.    36.222222           41.408869    103.0    0.0
3              perl.    41.333333           14.843629     54.0   25.0
4           neptune.     0.000000            0.000000      0.0    0.0
5             smurf.     0.000000            0.000000      0.0    0.0
6      guess_passwd.     2.716981           11.879811     60.0    0.0
7               pod.     0.000000            0.000000      0.0    0.0
8          teardrop.     0.000000            0.000000      0.0    0.0
9         portsweep.  1915.299038         7285.125159  42448.0    0.0
10          ipsweep.     0.034483            0.438439      7.0    0.0
11             land.     0.000000            0.000000      0.0    0.0
12        ftp_write.    32.375000           47.449033    134.0    0.0
13             back.     0.128915            1.110062     14.0    0.0
14             imap.     6.000000           14.174240     41.0    0.0
15            satan.     0.040277            0.522433     11.0    0.0
16              phf.     4.500000            5.744563     12.0    0.0
17             nmap.     0.000000            0.000000      0.0    0.0
18         multihop.   184.000000          253.851006    718.0    0.0
19      warezmaster.    15.050000           33.385271    156.0    0.0
20      warezclient.   615.257843         2207.694966  15168.0    0.0
21              spy.   318.000000           26.870058    337.0  299.0
22          rootkit.   100.800000          216.185003    708.0    0.0

### although we are doing it by only the duration feature we can pass a function to do it for any of the feature , lets pass a function for that 

In [ ]:
def feature_stats(sum_by_label,column_f):
  colum_stats_by_label=[(stat[0],stat[1].mean()[column_f],sqrt(stat[1].variance()[column_f]),stat[1].max()[column_f],stat[1].min()[column_f])for stat in sum_by_label]
  return pd.DataFrame(colum_stats_by_label,columns=['name','mean','standard deviation','max','min'])

In [ ]:
feature_stats(sum_by_label,0)## similarly we can get the summary by 1 ,2 ,3 , that are other features in the data 

name         mean  standard deviation      max    min
0            normal.   216.657322         1359.213469  58329.0    0.0
1   buffer_overflow.    91.700000           97.514685    321.0    0.0
2        loadmodule.    36.222222           41.408869    103.0    0.0
3              perl.    41.333333           14.843629     54.0   25.0
4           neptune.     0.000000            0.000000      0.0    0.0
5             smurf.     0.000000            0.000000      0.0    0.0
6      guess_passwd.     2.716981           11.879811     60.0    0.0
7               pod.     0.000000            0.000000      0.0    0.0
8          teardrop.     0.000000            0.000000      0.0    0.0
9         portsweep.  1915.299038         7285.125159  42448.0    0.0
10          ipsweep.     0.034483            0.438439      7.0    0.0
11             land.     0.000000            0.000000      0.0    0.0
12        ftp_write.    32.375000           47.449033    134.0    0.0
13             back.     0.128915            1.110062     14.0    0.0
14             imap.     6.000000           14.174240     41.0    0.0
15            satan.     0.040277            0.522433     11.0    0.0
16              phf.     4.500000            5.744563     12.0    0.0
17             nmap.     0.000000            0.000000      0.0    0.0
18         multihop.   184.000000          253.851006    718.0    0.0
19      warezmaster.    15.050000           33.385271    156.0    0.0
20      warezclient.   615.257843         2207.694966  15168.0    0.0
21              spy.   318.000000           26.870058    337.0  299.0
22          rootkit.   100.800000          216.185003    708.0    0.0

## we are going to start with the correlation , we have earlier cleared the num data , into the vector , we will do that , and then use the correlation for the rdd

In [ ]:
import numpy as np
def parse3(line):
  line_split=line.split(',')
  useful_indexes=[1,2,3,41]
  clear_line=[item for i, item in enumerate(line_split) if i not in useful_indexes]
  return (np.array([float(x) for x in clear_line]))

In [ ]:
df2=rdd.map(parse3)

In [ ]:
from pyspark.mllib.stat import Statistics as st
df_corr=st.corr(df2,method='spearman')


In [ ]:
import pandas as pd
col=["duration","src_bytes","dst_bytes","land","wrong_fragment",
             "urgent","hot","num_failed_logins","logged_in","num_compromised",
             "root_shell","su_attempted","num_root","num_file_creations",
             "num_shells","num_access_files","num_outbound_cmds",
             "is_hot_login","is_guest_login","count","srv_count","serror_rate",
             "srv_serror_rate","rerror_rate","srv_rerror_rate","same_srv_rate",
             "diff_srv_rate","srv_diff_host_rate","dst_host_count","dst_host_srv_count",
             "dst_host_same_srv_rate","dst_host_diff_srv_rate","dst_host_same_src_port_rate",
             "dst_host_srv_diff_host_rate","dst_host_serror_rate","dst_host_srv_serror_rate",
             "dst_host_rerror_rate","dst_host_srv_rerror_rate"]
corr_data=pd.DataFrame(df_corr,index=col,columns=col)

In [ ]:
corr_data

duration  ...  dst_host_srv_rerror_rate
duration                     1.000000  ...                 -0.013891
src_bytes                    0.014196  ...                 -0.300581
dst_bytes                    0.299189  ...                  0.001621
land                        -0.001068  ...                 -0.001820
wrong_fragment              -0.008025  ...                 -0.013666
urgent                       0.017884  ...                 -0.000776
hot                          0.108639  ...                  0.189142
num_failed_logins            0.014363  ...                  0.032151
logged_in                    0.159564  ...                  0.012979
num_compromised              0.010687  ...                  0.217859
root_shell                   0.040425  ...                  0.002156
su_attempted                 0.026012  ...                  0.001735
num_root                     0.013401  ...                 -0.005013
num_file_creations           0.061099  ...                 -0.001093
num_shells                   0.008633  ...                 -0.002024
num_access_files             0.019407  ...                 -0.004555
num_outbound_cmds                 NaN  ...                       NaN
is_hot_login                      NaN  ...                       NaN
is_guest_login               0.205607  ...                 -0.004289
count                       -0.259032  ...                 -0.256176
srv_count                   -0.250139  ...                 -0.308132
serror_rate                 -0.074211  ...                 -0.105434
srv_serror_rate             -0.073663  ...                 -0.124657
rerror_rate                 -0.025936  ...                  0.911623
srv_rerror_rate             -0.026420  ...                  0.914904
same_srv_rate                0.062291  ...                 -0.282914
diff_srv_rate               -0.050875  ...                  0.298904
srv_diff_host_rate           0.123621  ...                  0.024722
dst_host_count              -0.161107  ...                 -0.125273
dst_host_srv_count          -0.217167  ...                 -0.300787
dst_host_same_srv_rate      -0.211979  ...                 -0.264384
dst_host_diff_srv_rate       0.231644  ...                  0.271067
dst_host_same_src_port_rate -0.065202  ...                 -0.297100
dst_host_srv_diff_host_rate  0.100692  ...                  0.120767
dst_host_serror_rate        -0.056753  ...                 -0.096899
dst_host_srv_serror_rate    -0.057298  ...                 -0.110532
dst_host_rerror_rate        -0.007759  ...                  0.950963
dst_host_srv_rerror_rate    -0.013891  ...                  1.000000

[38 rows x 38 columns]

### we have the correlated data , we can use the highly correlated features in one frame , and slice the features based on the condition , better to check the true and false, than the values for a heavy data set, so convering it to boolean

In [ ]:
high_corr=(abs(corr_data)>.8) & (corr_data <1)
high_corr_index=(high_corr==True).any()
high_corr_names=high_corr_index[high_corr_index==True].index
high_corr.loc[high_corr_names,high_corr_names]

src_bytes  ...  dst_host_srv_rerror_rate
src_bytes                        False  ...                     False
dst_bytes                        False  ...                     False
hot                              False  ...                     False
logged_in                        False  ...                     False
num_compromised                  False  ...                     False
count                            False  ...                     False
srv_count                        False  ...                     False
serror_rate                      False  ...                     False
srv_serror_rate                  False  ...                     False
rerror_rate                      False  ...                      True
srv_rerror_rate                  False  ...                      True
same_srv_rate                    False  ...                     False
diff_srv_rate                    False  ...                     False
dst_host_count                   False  ...                     False
dst_host_srv_count               False  ...                     False
dst_host_same_srv_rate           False  ...                     False
dst_host_diff_srv_rate           False  ...                     False
dst_host_same_src_port_rate       True  ...                     False
dst_host_srv_diff_host_rate      False  ...                     False
dst_host_serror_rate             False  ...                     False
dst_host_srv_serror_rate         False  ...                     False
dst_host_rerror_rate             False  ...                      True
dst_host_srv_rerror_rate         False  ...                     False

[23 rows x 23 columns]

# MACHINE LEARNING

### now that we have all the analysis done , we will go ahead with the nmachine learning , we will use the logistic regg for this 


##### let us get the test data set for this 

In [ ]:
!wget -q http://kdd.ics.uci.edu/databases/kddcup99/corrected.gz

In [ ]:
test_data_file = "./corrected.gz"
test_raw_data = ap.textFile(test_data_file)

### now that we did miss some features like 1,2,3,41 , and printed the lable in the front we will do that as well here 

### also as we are doing log regg , we will like to lable our data as a binary , 0 means normal , attack as 1

In [ ]:
from pyspark.mllib.regression import LabeledPoint as lp

In [ ]:

def parse4(line):
  line_split=line.split(',')
  useful_indexes=[1,2,3,41]
  clear_line=[item for i, item in enumerate(line_split) if i not in useful_indexes]
  attack=1.0
  if line_split[41]=='normal.':
    attack=0.0
  return lp(attack,np.array([float(x) for x in clear_line]))

In [ ]:
train_rdd=rdd.map(parse4)

In [ ]:
train_rdd.take(5)

[LabeledPoint(0.0, [0.0,181.0,5450.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,9.0,9.0,1.0,0.0,0.11,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,239.0,486.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,19.0,19.0,1.0,0.0,0.05,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,235.0,1337.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,29.0,29.0,1.0,0.0,0.03,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,219.0,1337.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.0,6.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,39.0,39.0,1.0,0.0,0.03,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,217.0,2032.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.0,6.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,49.0,49.0,1.0,0.0,0.02,0.0,0.0,0.0,0.0,0.0])]

In [ ]:
test_rdd=test_raw_data.map(parse4)

In [ ]:
test_rdd.take(5)

[LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.01,0.0,0.0,0.0,0.0,0.0])]

## now that we have changed the lables to 0 and 1 , now we will train our log regg model

In [ ]:
from pyspark.mllib.classification import LogisticRegressionWithLBFGS as lg
from time import time

# Build the model
t0 = time()
logit_model = lg.train(train_rdd)
tt = time() - t0

print ("Classifier trained in {} seconds".format(round(tt,3)))

Classifier trained in 26.099 seconds


## now we will validate our model , using our test data , we will also check the lable and the predicted value for that model 

In [ ]:
labels_and_preds = test_rdd.map(lambda p: (p.label, logit_model.predict(p.features)))


In [ ]:
labels_and_preds.take(5)

[(0.0, 0), (0.0, 0), (0.0, 0), (1.0, 0), (1.0, 0)]

## we can see that we have 2 wrongly classfied values , that means we have a model ready , we will check the accuracy of our model


In [ ]:
t0 = time()
accu=round(labels_and_preds.filter(lambda x: x[0]==x[1]).count()/test_rdd.count(),4)
print('the accuracy of the model is:',accu*100,'%')
tt = time() - t0
print('the prediction was made in ',round(tt,3),'seconds')

the accuracy of the model is: 90.01 %
the prediction was made in  14.252 seconds


##Feature Engineering

### checking at the corre matrix above ,we will drop out the features : [1,2,3,25,27,35,38,40,41], and train our model again


In [ ]:
def parse_interaction_corr(line):
    line_split = line.split(",")
    # leave_out = [1,2,3,25,27,35,38,40,41]
    clean_line_split = line_split[0:1]+line_split[4:25]+line_split[26:27]+line_split[28:35]+line_split[36:38]+line_split[39:40]
    attack = 1.0
    if line_split[41]=='normal.':
        attack = 0.0
    return lp(attack, np.array([float(x) for x in clean_line_split]))

corr_reduced_training_data = rdd.map(parse_interaction_corr)
corr_reduced_test_data = test_raw_data.map(parse_interaction_corr)

In [ ]:
corr_reduced_test_data.take(5)

[LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0])]

## we are going to train the model with the selected features 

In [ ]:
# Build the model
t0 = time()
logit_model_2 = lg.train(corr_reduced_training_data)
tt = time() - t0

print ("Classifier trained in {} seconds".format(round(tt,3)))

Classifier trained in 25.324 seconds


In [ ]:
labels_pred_2= corr_reduced_test_data.map(lambda p: (p.label, logit_model_2.predict(p.features)))

In [ ]:
labels_pred_2.take(5)

[(0.0, 0), (0.0, 0), (0.0, 0), (1.0, 0), (1.0, 0)]

In [ ]:
t0=time()
accur2=labels_pred_2.filter(lambda x: x[0]==x[1]).count() / float(corr_reduced_test_data.count())
print('the accuracy is ',round(accur2*100,4))
tt=time()-t0
print('the prediction made in ',round(tt,3),'seconds')

the accuracy is  81.5815
the prediction made in  11.349 seconds


## there was a huge loss of accuracy , and very low amount of time reduced , dont think that giving up such features will be helpfull for accuracy , let us try with some other modelling method

# Dec Tree

In [ ]:
rdd.take(1)

['0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.']

#### as we have text field in the data , we will change that into the numerical by separating it from the RDD , and then use the map function to reverse back the numerical values

In [ ]:
csv_data = rdd.map(lambda x: x.split(","))
test_csv_data = test_raw_data.map(lambda x: x.split(","))

protocols = csv_data.map(lambda x: x[1]).distinct().collect()
services = csv_data.map(lambda x: x[2]).distinct().collect()
flags = csv_data.map(lambda x: x[3]).distinct().collect()

In [ ]:
def create_labeled_point(line_split):
    # leave_out = [41]
    clean_line_split = line_split[0:41]
    
    # convert protocol to numeric categorical variable
    try: 
        clean_line_split[1] = protocols.index(clean_line_split[1])
    except:
        clean_line_split[1] = len(protocols)
        
    # convert service to numeric categorical variable
    try:
        clean_line_split[2] = services.index(clean_line_split[2])
    except:
        clean_line_split[2] = len(services)
    
    # convert flag to numeric categorical variable
    try:
        clean_line_split[3] = flags.index(clean_line_split[3])
    except:
        clean_line_split[3] = len(flags)
    
    # convert label to binary label
    attack = 1.0
    if line_split[41]=='normal.':
        attack = 0.0
        
    return lp(attack, np.array([float(x) for x in clean_line_split]))

training_data = csv_data.map(create_labeled_point)
test_data = test_csv_data.map(create_labeled_point)

In [ ]:
training_data.take(5)

[LabeledPoint(0.0, [0.0,0.0,0.0,0.0,181.0,5450.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,9.0,9.0,1.0,0.0,0.11,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,0.0,0.0,0.0,239.0,486.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,19.0,19.0,1.0,0.0,0.05,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,0.0,0.0,0.0,235.0,1337.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,8.0,8.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,29.0,29.0,1.0,0.0,0.03,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,0.0,0.0,0.0,219.0,1337.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.0,6.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,39.0,39.0,1.0,0.0,0.03,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,0.0,0.0,0.0,217.0,2032.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,6.0,6.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,49.0,49.0,1.0,0.0,0.02,0.0,0.0,0.0,0.0,0.0])]

In [ ]:
test_data.take(5)

[LabeledPoint(0.0, [0.0,1.0,11.0,0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,1.0,11.0,0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(0.0, [0.0,1.0,11.0,0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,1.0,11.0,0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.0,0.0,0.0,0.0,0.0,0.0]),
 LabeledPoint(1.0, [0.0,1.0,11.0,0.0,105.0,146.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,2.0,2.0,0.0,0.0,0.0,0.0,1.0,0.0,0.0,255.0,254.0,1.0,0.01,0.01,0.0,0.0,0.0,0.0,0.0])]

In [ ]:
from pyspark.mllib.tree import DecisionTree, DecisionTreeModel
from time import time

# Build the model
t0 = time()
tree_model = DecisionTree.trainClassifier(training_data, numClasses=2, 
                                          categoricalFeaturesInfo={1: len(protocols), 2: len(services), 3: len(flags)},
                                          impurity='entropy', maxDepth=4, maxBins=100)
tt = time() - t0

print ("Classifier trained in {} seconds".format(round(tt,3)))

Classifier trained in 14.614 seconds


In [ ]:
predictions = tree_model.predict(test_data.map(lambda p: p.features))
labels_and_preds3 = test_data.map(lambda p: p.label).zip(predictions)

In [ ]:
t0 = time()
test_accuracy = labels_and_preds3.filter(lambda x:x[0]==x[1]).count() / float(test_data.count())
tt = time() - t0

print ("Prediction made in {} seconds. Test accuracy is {}".format(round(tt,3), round(test_accuracy,4)))

Prediction made in 16.929 seconds. Test accuracy is 0.9219


### we have a good time and good accuracy than the log regg

##### let us see the dec tree that was created 

In [ ]:
print ("Learned classification tree model:")
print (tree_model.toDebugString())

Learned classification tree model:
DecisionTreeModel classifier of depth 4 with 23 nodes
  If (feature 22 <= 53.5)
   If (feature 4 <= 28.5)
    If (feature 2 in {0.0,5.0,56.0,42.0,24.0,37.0,25.0,14.0,20.0,46.0,57.0,29.0,1.0,6.0,60.0,28.0,38.0,21.0,33.0,9.0,53.0,13.0,41.0,2.0,32.0,34.0,45.0,17.0,22.0,44.0,27.0,12.0,54.0,49.0,39.0,3.0,35.0,48.0,63.0,18.0,50.0,16.0,31.0,43.0,40.0,55.0,23.0,58.0,36.0,30.0,51.0,19.0,4.0,47.0,15.0})
     If (feature 24 <= 0.5800000000000001)
      Predict: 0.0
     Else (feature 24 > 0.5800000000000001)
      Predict: 1.0
    Else (feature 2 not in {0.0,5.0,56.0,42.0,24.0,37.0,25.0,14.0,20.0,46.0,57.0,29.0,1.0,6.0,60.0,28.0,38.0,21.0,33.0,9.0,53.0,13.0,41.0,2.0,32.0,34.0,45.0,17.0,22.0,44.0,27.0,12.0,54.0,49.0,39.0,3.0,35.0,48.0,63.0,18.0,50.0,16.0,31.0,43.0,40.0,55.0,23.0,58.0,36.0,30.0,51.0,19.0,4.0,47.0,15.0})
     Predict: 1.0
   Else (feature 4 > 28.5)
    If (feature 9 <= 0.5)
     Predict: 0.0
    Else (feature 9 > 0.5)
     If (feature 4 <= 1092.0)


### let us build model , leaving the 42 feature and transforming the dependent variable in 1 and 0 

In [ ]:
def create_labeled_point_minimal(line_split):
    # leave_out = [41]
    clean_line_split = line_split[3:4] + line_split[5:6] + line_split[22:23]
    
    # convert flag to numeric categorical variable
    try:
        clean_line_split[0] = flags.index(clean_line_split[0])
    except:
        clean_line_split[0] = len(flags)
    
    # convert label to binary label
    attack = 1.0
    if line_split[41]=='normal.':
        attack = 0.0
        
    return lp(attack, np.array([float(x) for x in clean_line_split]))

training_data_minimal = csv_data.map(create_labeled_point_minimal)
test_data_minimal = test_csv_data.map(create_labeled_point_minimal)
# same as we did in log reg , we will leave out the features

In [ ]:
training_data_minimal.take(2)

[LabeledPoint(0.0, [0.0,5450.0,8.0]), LabeledPoint(0.0, [0.0,486.0,8.0])]

### we will build the model 

In [ ]:
t0 = time()
tree_model_minimal = DecisionTree.trainClassifier(training_data_minimal, numClasses=2, 
                                          categoricalFeaturesInfo={0: len(flags)},
                                          impurity='gini', maxDepth=3, maxBins=32)
tt = time() - t0

print ("Classifier trained in {} seconds".format(round(tt,3)))

Classifier trained in 9.946 seconds


In [ ]:
predictions_minimal = tree_model_minimal.predict(test_data_minimal.map(lambda p: p.features))
labels_and_preds_minimal = test_data_minimal.map(lambda p: p.label).zip(predictions_minimal)

### checking the accuracy and comparing it with the log regg model

In [ ]:
t0 = time()
test_accuracy = labels_and_preds_minimal.filter(lambda x:x[0]==x[1]).count() / float(test_data_minimal.count())
tt = time() - t0

print ("Prediction made in {} seconds. Test accuracy is {} %".format(round(tt,3), round(test_accuracy*100,4)))

Prediction made in 9.795 seconds. Test accuracy is 91.6818 %


## let us compare it with the log regg model

In [ ]:
t0=time()
accur2=labels_pred_2.filter(lambda x: x[0]==x[1]).count() / float(corr_reduced_test_data.count())
print('the accuracy is ',round(accur2*100,4),'%')
tt=time()-t0
print('the prediction made in ',round(tt,3),'seconds')

the accuracy is  81.5815 %
the prediction made in  11.435 seconds


## we can clearly see that the dec tree model is out performing in both the full features data and slpitted features model , with the good time as compared to log regg for full data and the splitted features model

# SQL with SPARK

### for SQL we will run the pyspark.sql and import the sql context , with that we will begin our sql journey 

In [ ]:
from pyspark.sql import SQLContext
sqlContext = SQLContext(ap)

/usr/local/lib/python3.7/dist-packages/pyspark/sql/context.py:79: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  FutureWarning


### now we will create a SQL data frame from our existing RDD,by passing the key value pair for the RDD

In [ ]:
from pyspark.sql import Row

csv_data = rdd.map(lambda l: l.split(","))
row_data = csv_data.map(lambda p: Row(
    duration=int(p[0]), 
    protocol_type=p[1],
    service=p[2],
    flag=p[3],
    src_bytes=int(p[4]),
    dst_bytes=int(p[5])
    )
)

### now that we have our RDD prepared we will put the data init 

In [ ]:
interactions_df = sqlContext.createDataFrame(row_data)
interactions_df.registerTempTable("interactions")

/usr/local/lib/python3.7/dist-packages/pyspark/sql/dataframe.py:140: FutureWarning: Deprecated in 2.0, use createOrReplaceTempView instead.
  FutureWarning


## now we can run our SQL queries 

In [ ]:
rd=sqlContext.sql("select * from interactions").show()

+--------+-------------+-------+----+---------+---------+
|duration|protocol_type|service|flag|src_bytes|dst_bytes|
+--------+-------------+-------+----+---------+---------+
|       0|          tcp|   http|  SF|      181|     5450|
|       0|          tcp|   http|  SF|      239|      486|
|       0|          tcp|   http|  SF|      235|     1337|
|       0|          tcp|   http|  SF|      219|     1337|
|       0|          tcp|   http|  SF|      217|     2032|
|       0|          tcp|   http|  SF|      217|     2032|
|       0|          tcp|   http|  SF|      212|     1940|
|       0|          tcp|   http|  SF|      159|     4087|
|       0|          tcp|   http|  SF|      210|      151|
|       0|          tcp|   http|  SF|      212|      786|
|       0|          tcp|   http|  SF|      210|      624|
|       0|          tcp|   http|  SF|      177|     1985|
|       0|          tcp|   http|  SF|      222|      773|
|       0|          tcp|   http|  SF|      256|     1169|
|       0|    

### we are creating an inteaction for the sql , and that is the reason we assign a variable to the rdd path , and then .show() shows us the value of the written sql code

In [ ]:
sqlContext.sql("select duration,dst_bytes from interactions where protocol_type='tcp' AND duration>1000 AND dst_bytes=0  ").show()

+--------+---------+
|duration|dst_bytes|
+--------+---------+
|    5057|        0|
|    5059|        0|
|    5051|        0|
|    5056|        0|
|    5051|        0|
|    5039|        0|
|    5062|        0|
|    5041|        0|
|    5056|        0|
|    5064|        0|
|    5043|        0|
|    5061|        0|
|    5049|        0|
|    5061|        0|
|    5048|        0|
|    5047|        0|
|    5044|        0|
|    5063|        0|
|    5068|        0|
|    5062|        0|
+--------+---------+
only showing top 20 rows



## the results given out as itself RDD's as written above with the explained reasons

In [ ]:
tcp_in=sqlContext.sql("select duration,dst_bytes from interactions where protocol_type='tcp' AND duration>1000 AND dst_bytes=0  ")

In [ ]:
tcp_in_df=tcp_in.map(lambda p:'duration: {} ,dest_bytes: {}'.format(p.duration,p.dst_bytes))
for ti in tcp_in.collect():
  print(tcp_in)

AttributeError: ignored

### checking to solve the error

In [ ]:
tcp_interactions = sqlContext.sql("""
    SELECT duration, dst_bytes FROM interactions WHERE protocol_type = 'tcp' AND duration > 1000 AND dst_bytes = 0
""")
tcp_interactions.show()

+--------+---------+
|duration|dst_bytes|
+--------+---------+
|    5057|        0|
|    5059|        0|
|    5051|        0|
|    5056|        0|
|    5051|        0|
|    5039|        0|
|    5062|        0|
|    5041|        0|
|    5056|        0|
|    5064|        0|
|    5043|        0|
|    5061|        0|
|    5049|        0|
|    5061|        0|
|    5048|        0|
|    5047|        0|
|    5044|        0|
|    5063|        0|
|    5068|        0|
|    5062|        0|
+--------+---------+
only showing top 20 rows



In [ ]:
tcp_interactions_out = tcp_interactions.rdd.map(lambda p: "Duration: {}, Dest. bytes: {}".format(p.duration, p.dst_bytes))
for ti_out in tcp_interactions_out.collect():
  print(ti_out)

Duration: 5057, Dest. bytes: 0
Duration: 5059, Dest. bytes: 0
Duration: 5051, Dest. bytes: 0
Duration: 5056, Dest. bytes: 0
Duration: 5051, Dest. bytes: 0
Duration: 5039, Dest. bytes: 0
Duration: 5062, Dest. bytes: 0
Duration: 5041, Dest. bytes: 0
Duration: 5056, Dest. bytes: 0
Duration: 5064, Dest. bytes: 0
Duration: 5043, Dest. bytes: 0
Duration: 5061, Dest. bytes: 0
Duration: 5049, Dest. bytes: 0
Duration: 5061, Dest. bytes: 0
Duration: 5048, Dest. bytes: 0
Duration: 5047, Dest. bytes: 0
Duration: 5044, Dest. bytes: 0
Duration: 5063, Dest. bytes: 0
Duration: 5068, Dest. bytes: 0
Duration: 5062, Dest. bytes: 0
Duration: 5046, Dest. bytes: 0
Duration: 5052, Dest. bytes: 0
Duration: 5044, Dest. bytes: 0
Duration: 5054, Dest. bytes: 0
Duration: 5039, Dest. bytes: 0
Duration: 5058, Dest. bytes: 0
Duration: 5051, Dest. bytes: 0
Duration: 5032, Dest. bytes: 0
Duration: 5063, Dest. bytes: 0
Duration: 5040, Dest. bytes: 0
Duration: 5051, Dest. bytes: 0
Duration: 5066, Dest. bytes: 0
Duration

### was getting an error that Dataframe has no attribute map , and that is true , so I found out that we need to use .rdd in that to change the dataframe to rdd and then print the dataframe as we wanted

In [ ]:
tcp_interactions.printSchema()

root
 |-- duration: long (nullable = true)
 |-- dst_bytes: long (nullable = true)



In [ ]:
interactions_df.printSchema()

root
 |-- duration: long (nullable = true)
 |-- protocol_type: string (nullable = true)
 |-- service: string (nullable = true)
 |-- flag: string (nullable = true)
 |-- src_bytes: long (nullable = true)
 |-- dst_bytes: long (nullable = true)



### these are the dataframes that has the feature for to check the column types , which does not work with RDDS, let us work more with dataframes

In [ ]:
interactions_df.select('protocol_type','duration').show()

+-------------+--------+
|protocol_type|duration|
+-------------+--------+
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
|          tcp|       0|
+-------------+--------+
only showing top 20 rows



##let us check the count of the prototypes , with different entries

In [ ]:
interactions_df.select('protocol_type').groupBy('protocol_type').count().show()

+-------------+------+
|protocol_type| count|
+-------------+------+
|          tcp|190065|
|          udp| 20354|
|         icmp|283602|
+-------------+------+



## as per the subject matter , let us find out the entries with duration more than 1 sec and no data trf from that entry 

In [ ]:
t0 = time()
interactions_df.select("protocol_type").filter(interactions_df.duration>=1).filter(interactions_df.dst_bytes==0).groupBy("protocol_type").count().show()
tt = time() - t0

print ("Query performed in {} seconds".format(round(tt,3)))

+-------------+-----+
|protocol_type|count|
+-------------+-----+
|          tcp|  342|
+-------------+-----+

Query performed in 6.028 seconds


### we will do some EDA on the data , we will create an Dataframe for the rdd , and then move the data into it , then we will do some analysis that we did above

In [ ]:
def get_label_type(label):
    if label!="normal.":
        return "attack"
    else:
        return "normal"
    
row_labeled_data = csv_data.map(lambda p: Row(
    duration=int(p[0]), 
    protocol_type=p[1],
    service=p[2],
    flag=p[3],
    src_bytes=int(p[4]),
    dst_bytes=int(p[5]),
    label=get_label_type(p[41])
    )
)
interactions_labeled_df = sqlContext.createDataFrame(row_labeled_data)

##let us check the counts for normal and attack that we separated in the function above 

In [ ]:
t0 = time()
interactions_labeled_df.select("label").groupBy("label").count().show()
tt = time() - t0

print ("Query performed in {} seconds".format(round(tt,3)))

+------+------+
| label| count|
+------+------+
|normal| 97278|
|attack|396743|
+------+------+

Query performed in 6.545 seconds


## the same count that we got above , so this is also a good way to do the eda on the data set , either using Rdd or the dataframe 

In [ ]:
t0 = time()
interactions_labeled_df.select("label", "protocol_type").groupBy("label", "protocol_type").count().show()
tt = time() - t0

print ("Query performed in {} seconds".format(round(tt,3)))

+------+-------------+------+
| label|protocol_type| count|
+------+-------------+------+
|normal|          udp| 19177|
|normal|         icmp|  1288|
|normal|          tcp| 76813|
|attack|         icmp|282314|
|attack|          tcp|113252|
|attack|          udp|  1177|
+------+-------------+------+

Query performed in 6.392 seconds


### we did count for normal and attack for diff prototype_type 

## let us see the different features and count  for when the attacker was present for 0 seconds only 


In [ ]:
t0 = time()
interactions_labeled_df.select("label", "protocol_type", "dst_bytes").filter(interactions_labeled_df.dst_bytes==0).groupBy("label", "protocol_type").count().show()
tt = time() - t0

print ("Query performed in {} seconds".format(round(tt,3)))

+------+-------------+------+
| label|protocol_type| count|
+------+-------------+------+
|normal|          udp|  3594|
|normal|         icmp|  1288|
|normal|          tcp|  9313|
|attack|         icmp|282314|
|attack|          tcp|110583|
|attack|          udp|  1166|
+------+-------------+------+

Query performed in 6.227 seconds


## here we can see that mostly the attackers who stay for 0 seconds or leave due to any interuption are using icmp or tcp mostly 

## Here we some of the things I tried ,
# Please let me know in the comments if you have any query for me 